File: typescript/servers/darena-solutions/fhir-crud/mcp-tools/find-patient-id-tool.ts
import { z } from "zod";
import { getFhirContext, getPatientIdIfContextExists } from "../fhir-utilities";
import { createTextResponse } from "../mcp-utilities";
import axios from "axios";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp";
import { Request } from "express";
import { IMcpTool } from "../IMcpTool";

class FindPatientIdTool implements IMcpTool {
  registerTool(server: McpServer, req: Request) {
    server.tool(
      "find_patient_id",
      "Finds an id of the patient. It is not required to supply both first name and " +
        "last name, only one of the parameters is sufficient. If patient context already " +
        "exists, then do not supply first name and last name unless the user explicitly " +
        "provides this information.",
      {
        firstName: z.string().optional().describe("The patient's first name"),
        lastName: z.string().optional().describe("The patient's last name"),
      },
      async ({ firstName, lastName }, extra) => {
        const fhirContext = getFhirContext(req);
        if (!fhirContext) {
          return createTextResponse(
            "A FHIR server url or token was not provided in the http context.",
            { isError: true }
          );
        }

        const patientIdContext = getPatientIdIfContextExists(req);
        if (patientIdContext) {
          return createTextResponse(patientIdContext);
        }

        if (!firstName && !lastName) {
          return createTextResponse(
            "No patient context found, thus the first name or last name is required.",
            { isError: true }
          );
        }

        const patientSearcher = async (
          searchFirstName: string | null | undefined,
          searchLastName: string | null | undefined
        ): Promise<{ id: string }[] | null> => {
          const searchParameters: string[] = [];
          if (searchFirstName) {
            searchParameters.push(`given=${searchFirstName}`);
          }

          if (searchLastName) {
            searchParameters.push(`family=${searchLastName}`);
          }

          const url = `${fhirContext.url}/Patient?${searchParameters.join(
            "&"
          )}`;
          const { data } = await axios.get(url, {
            headers: {
              Authorization: `Bearer ${fhirContext.token}`,
            },
          });

          return data?.entry.length
            ? (data.entry as { resource: { id: string } }[]).map(
                (x) => x.resource
              )
            : null;
        };

        try {
          let searchResponse = await patientSearcher(firstName, lastName);
          if (!searchResponse) {
            searchResponse = await patientSearcher(lastName, firstName);
          }

          if (!searchResponse || searchResponse.length === 0) {
            return createTextResponse("The patient could not be found.", {
              isError: true,
            });
          }

          if (searchResponse.length > 1) {
            return createTextResponse(
              "More than one patient found. Need more details",
              { isError: true }
            );
          }

          return createTextResponse(searchResponse[0].id);
        } catch (error) {
          return createTextResponse(
            "An unexpected error occurred while attempting to retrieve the patient's id.",
            { isError: true }
          );
        }
      }
    );
  }
}

export const FindPatientIdToolInstance = new FindPatientIdTool();

File: dotnet/servers/darena-solutions/DarenaHealth.Community.McpServers.DarenaSolutions.FhirCrud/Tools/FindPatientIdTool.cs
using DarenaHealth.Community.Mcp.Core;
using DarenaHealth.Community.Mcp.Core.Extensions;
using DarenaHealth.Community.Mcp.Core.Models;
using Hl7.Fhir.Model;
using Hl7.Fhir.Rest;
using ModelContextProtocol.Protocol;
using ModelContextProtocol.Server;

namespace DarenaHealth.Community.McpServers.DarenaSolutions.FhirCrud.Tools;

public class FindPatientIdTool : IMcpTool
{
    private const string FirstNameParameter = "firstName";
    private const string LastNameParameter = "lastName";

    public string Name { get; } = "find_patient_id";

    public string? Description { get; } =
        "Finds an id of the patient. It is not required to supply both first name and last name, only one of the parameters "
        + "is sufficient. If patient context already exists, then do not supply first name and last name unless the user "
        + "explicitly provides this information.";

    public List<McpToolArgument> Arguments { get; } =
        [
            new McpToolArgument
            {
                Type = "string",
                Name = FirstNameParameter,
                Description = "The patient's first name",
                IsRequired = false,
            },
            new McpToolArgument
            {
                Type = "string",
                Name = LastNameParameter,
                Description = "The patient's last name",
                IsRequired = false,
            },
        ];

    public List<string> FhirScopes { get; } = ["patient/Patient.read"];

    public async Task<CallToolResponse> HandleAsync(
        HttpContext httpContext,
        IMcpServer mcpServer,
        CallToolRequestParams context
    )
    {
        var fhirContext = httpContext.GetFhirContext();
        if (fhirContext is null)
        {
            return McpToolUtilities.CreateTextToolResponse(
                "A FHIR server url or token was not provided in the http context.",
                isError: true
            );
        }

        var patientIdContext = httpContext.GetPatientIdIfContextExists();
        if (!string.IsNullOrWhiteSpace(patientIdContext))
        {
            return McpToolUtilities.CreateTextToolResponse(patientIdContext);
        }

        if (context.Arguments is null)
        {
            return McpToolUtilities.CreateTextToolResponse(
                "No patient context found, thus the first name or last name is required.",
                isError: true
            );
        }

        var firstName = context.GetArgumentValueOrNull(FirstNameParameter);
        var lastName = context.GetArgumentValueOrNull(LastNameParameter);

        if (string.IsNullOrWhiteSpace(firstName) && string.IsNullOrWhiteSpace(lastName))
        {
            return McpToolUtilities.CreateTextToolResponse(
                "No patient context found, thus the first name or last name is required.",
                isError: true
            );
        }

        var fhirSettings = new FhirClientSettings { PreferredFormat = ResourceFormat.Json };
        var fhirClient = new FhirClient(
            fhirContext.Url,
            fhirSettings,
            new FhirClientAuthMessageHandler(fhirContext.Token)
        );

        var searchParams = new SearchParams();
        if (!string.IsNullOrWhiteSpace(firstName))
        {
            searchParams.Where($"given={firstName}");
        }

        if (!string.IsNullOrWhiteSpace(lastName))
        {
            searchParams.Where($"family={lastName}");
        }

        try
        {
            var response = await fhirClient.SearchAsync<Patient>(searchParams);
            if (response?.Entry is not { Count: > 0 })
            {
                searchParams = new SearchParams();
                if (!string.IsNullOrWhiteSpace(lastName))
                {
                    searchParams.Where($"given={lastName}");
                }

                if (!string.IsNullOrWhiteSpace(firstName))
                {
                    searchParams.Where($"family={firstName}");
                }

                response = await fhirClient.SearchAsync<Patient>(searchParams);
                if (response?.Entry is not { Count: > 0 })
                {
                    return McpToolUtilities.CreateTextToolResponse(
                        "The patient could not be found.",
                        isError: true
                    );
                }
            }

            if (response.Entry.Count > 1)
            {
                return McpToolUtilities.CreateTextToolResponse(
                    "More than one patient found. Need more details",
                    isError: true
                );
            }

            return McpToolUtilities.CreateTextToolResponse(response.Entry[0].Resource.Id);
        }
        catch (Exception e)
        {
            return McpToolUtilities.CreateTextToolResponse(
                $"An exception occurred with message: {e.Message}",
                isError: true
            );
        }
    }
}
